src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,514 @@
+    case state::more: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return more(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfcap: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfcap(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfreq: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfreq(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::complete:
+    case state::failed:
+        break;
+    }
+    vlog(seclog.info, "gss {} authenticate failed", _state);
+    _state = state::failed;
+    co_return errc::invalid_scram_state;
+}
+
+result<void>
+gssapi_authenticator::init(ss::sstring principal, ss::sstring keytab) {
+    OM_uint32 minor_status{};
+    OM_uint32 major_status{};
+
+    gss::buffer_view service_name{principal};
+    gss::name server_name{};
+
+    major_status = ::gss_import_name(
+      &minor_status, &service_name, GSS_C_NT_HOSTBASED_SERVICE, &server_name);
+
+    vlog(
+      seclog.trace,
+      "gss {} import name {}:{} - {}",
+      _state,
+      major_status,
+      minor_status,
+      principal);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("import name", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+
+    gss_key_value_element_desc elem{.key = "keytab", .value = keytab.c_str()};
+    gss_key_value_set_desc store{.count = 1, .elements = &elem};
+
+    major_status = ::gss_acquire_cred_from(
+      &minor_status,
+      server_name,
+      0,
+      GSS_C_NO_OID_SET,
+      GSS_C_ACCEPT,
+      &store,
+      &_server_creds,
+      NULL,
+      NULL);
+
+    vlog(
+      seclog.info,
+      "gss {} acquire cred {}:{} - {}",
+      _state,
+      major_status,
+      minor_status,
+      keytab);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("gss acquire cred", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+
+    gss::buffer token;
+    major_status = gss_export_cred(&minor_status, _server_creds, &token);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("gss acquire cred", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+    vlog(
+      seclog.trace,
+      "gss {} gss_export_cred: {}",
+      _state,
+      hex(bytes_view{token}));
```

Comment:
i don't know exactly what all the state is that's being logged, but i recall we need to be careful about logging anything that might be considered sensitive.

Reply:
Replaced by length